android/src/com/mapswithme/maps/downloader/DownloaderAdapter.java

```diff
@@ -364,68 +311,98 @@ public void run()
 
     private void processLongClick()
     {
-      List<MenuItem> items = new ArrayList<>();
+      ArrayList<MenuBottomSheetItem> items = getMenuItems();
+      if (items.isEmpty())
+        return;
 
-      switch (mItem.status)
-      {
-      case CountryItem.STATUS_DOWNLOADABLE:
-        items.add(MenuItem.DOWNLOAD);
-        break;
+      new MenuBottomSheetFragment(mItem.name, items)
+          .show(mFragment.getParentFragmentManager(), "downloaderBottomSheet");
+    }
 
-      case CountryItem.STATUS_UPDATABLE:
-        items.add(MenuItem.UPDATE);
-        // No break
+    private MenuBottomSheetItem getDownloadMenuItem()
+    {
+      return new MenuBottomSheetItem(
+          R.string.downloader_download_map,
+          R.drawable.ic_download,
+          () -> onDownloadActionSelected(mItem, DownloaderAdapter.this));
+    }
 
-      case CountryItem.STATUS_DONE:
-        if (!mItem.isExpandable())
-          items.add(MenuItem.EXPLORE);
+    private MenuBottomSheetItem getUpdateMenuItem()
+    {
+      return new MenuBottomSheetItem(
+          R.string.downloader_update_map,
+          R.drawable.ic_update,
+          () -> onUpdateActionSelected(mItem, DownloaderAdapter.this));
+    }
 
-        items.add(MenuItem.DELETE);
-        break;
+    private MenuBottomSheetItem getExploreMenuItem()
+    {
+      return new MenuBottomSheetItem(
+          R.string.zoom_to_country,
+          R.drawable.ic_explore,
+          () -> onExploreActionSelected(mItem, DownloaderAdapter.this));
+    }
 
-      case CountryItem.STATUS_FAILED:
-        items.add(MenuItem.CANCEL);
+    private MenuBottomSheetItem getDeleteMenuItem()
+    {
+      return new MenuBottomSheetItem(
+          R.string.delete,
+          R.drawable.ic_delete,
+          () -> onDeleteActionSelected(mItem, DownloaderAdapter.this));
+    }
 
-        if (mItem.present)
-        {
-          items.add(MenuItem.DELETE);
-          items.add(MenuItem.EXPLORE);
-        }
-        break;
+    private MenuBottomSheetItem getCancelMenuItem()
+    {
+      return new MenuBottomSheetItem(
+          R.string.cancel,
+          R.drawable.ic_cancel,
+          () -> onCancelActionSelected(mItem, DownloaderAdapter.this));
+    }
 
-      case CountryItem.STATUS_PROGRESS:
-      case CountryItem.STATUS_APPLYING:
-      case CountryItem.STATUS_ENQUEUED:
-        items.add(MenuItem.CANCEL);
+    private ArrayList<MenuBottomSheetItem> getMenuItems()
+    {
+      ArrayList<MenuBottomSheetItem> items = new ArrayList<>();
+      switch (mItem.status)
+      {
+        case CountryItem.STATUS_DOWNLOADABLE:
+          items.add(getDownloadMenuItem());
+          break;
 
-        if (mItem.present)
-          items.add(MenuItem.EXPLORE);
-        break;
+        case CountryItem.STATUS_UPDATABLE:
+          items.add(getUpdateMenuItem());
+          // No break
```

Comment:
nit: Is `// Fallthrough` better here and in other similar places? it is usually supported by linters, at least in C++.

Reply:
I don't know I have never used Java linters